jdaviz/configs/cubeviz/plugins/viewers.py

```diff
@@ -67,25 +81,44 @@ def slice_component_label(self):
             raise ValueError("slice plugin must be activated to access slice_component_label")
         return slice_plg._obj.slice_indicator_viewers[0].slice_component_label
 
-    @property
+    @cached_property
     def slice_values(self):
-        # TODO: make a cached property and invalidate cache on add/remove data
         # TODO: add support for multiple cubes (but then slice selection needs to be more complex)
         # if slice_index is 0, then we want the equivalent of [:, 0, 0]
         # if slice_index is 1, then we want the equivalent of [0, :, 0]
         # if slice_index is 2, then we want the equivalent of [0, 0, :]
         take_inds = [2, 1, 0]
         take_inds.remove(self.slice_index)
         for layer in self.layers:
+            world_comp_ids = layer.layer.data.world_component_ids
+            if self.slice_index >= len(world_comp_ids):
+                # Case where 2D image is loaded in image viewer
+                continue
             try:
-                data_obj = layer.layer.data.get_component(self.slice_component_label).data
+                # Retrieve display units
+                display_spectral_units = self.jdaviz_app._get_display_unit('spectral')
+
+                # Retrieve layer data and units using the slice index of the world components ids
+                data_obj = layer.layer.data.get_component(world_comp_ids[self.slice_index]).data
+                data_units = layer.layer.data.get_component(world_comp_ids[self.slice_index]).units
+
+                # Find the spectral axis
+                data_spec_axis = np.asarray(data_obj.take(0, take_inds[0]).take(0, take_inds[1]),  # noqa
+                                            dtype=float)
+                if display_spectral_units and display_spectral_units != data_units:
+                    converted_axis = (data_spec_axis * u.Unit(data_units)).to_value(
+                        display_spectral_units,
+                        equivalencies=u.spectral()
+                    )
+                else:
+                    converted_axis = data_spec_axis
```

Comment:
same thing here, it would be nice if "spectral"/"spec" isn't hardcoded in this method so downstream apps can use this mixin for non-spectral data